lib/services/access_checker.go

```diff
@@ -319,30 +321,67 @@ func (a *accessChecker) CheckAccess(r AccessCheckable, state AccessState, matche
 // GetKubeResources returns the allowed and denied Kubernetes Resources configured
 // for a user.
 func (a *accessChecker) GetKubeResources(cluster types.KubeCluster) (allowed, denied []types.KubernetesResource) {
-	if len(a.info.AllowedResourceIDs) > 0 {
-		for _, r := range a.info.AllowedResourceIDs {
-			if r.Kind == types.KindKubePod && r.Name == cluster.GetName() && r.ClusterName == a.localCluster {
-				splitted := strings.SplitN(r.SubResourceName, "/", 3)
-				// This condition should never happen since SubResourceName is validated
-				// but it's better to validate it.
-				if len(splitted) != 2 {
-					continue
-				}
-				allowed = append(allowed,
-					types.KubernetesResource{
-						Kind:      r.Kind,
-						Namespace: splitted[0],
-						Name:      splitted[1],
-					},
-				)
+	if len(a.info.AllowedResourceIDs) == 0 {
+		return a.RoleSet.GetKubeResources(cluster)
+	}
+
+	rolesAllowed, rolesDenied := a.RoleSet.GetKubeResources(cluster)
+	// Allways append the denied resources from the roles. This is because
+	// the denied resources from the roles take precedence over the allowed
+	// resources from the certificate.
+	denied = append(denied, rolesDenied...)
+	for _, r := range a.info.AllowedResourceIDs {
+		if r.Name != cluster.GetName() || r.ClusterName != a.localCluster {
+			continue
+		}
+		switch {
+		case slices.Contains(types.KubernetesResourcesKinds, r.Kind):
+			splitted := strings.SplitN(r.SubResourceName, "/", 3)
+			// This condition should never happen since SubResourceName is validated
+			// but it's better to validate it.
+			if len(splitted) != 2 {
+				continue
+			}
+
+			r := types.KubernetesResource{
+				Kind:      r.Kind,
+				Namespace: splitted[0],
+				Name:      splitted[1],
+			}
+
+			if matchKubernetesResource(r, rolesAllowed, rolesDenied) == nil {
+				allowed = append(allowed, r)
 			}
+		case r.Kind == types.KindKubernetesCluster:
+			// When a user has access to a Kubernetes cluster through Resource Access request,
+			// he has access to all resources in that cluster that he has access to through his roles.
+			// In that case, we append the allowed and denied resources from the roles.
+			allowed = append(allowed, rolesAllowed...)
```

Comment:
nit: skip some iterations, avoid the copy, possibly reduce the number of filters that need to be compared for the list pods endpoint
```suggestion
			// In that case, we return the allowed and denied resources from the roles, which are
			// guaranteed to be at least as permissive as any allowed subresources.
			return rolesAllowed, rolesDenied
```